handle styles

little-dude opened the pull request.

Commit:
add support for `set_style` rpc call from core

Commit:
get rid of so-called "fake cursors"

These fake cursors were initially a workaround before we had proper
cursor handling. I kept them around in prevision of multi-cursor support
in xi-core. But they will make style handling complicated because they
mess with the utf-8 codepoint indices in the lines, so let's get rid of
them and revisit multi-cursor when it actually lands. We'll probably
have to handle that the same way we currently handle the "real" cursor.

Commit:
get selection style working

selection is considered a style. It has the reserved style id 0. We
don't check the style id for the moment and always assume that the style
is a selection.

Commit:
update clap

Commit:
update README

Commit:
fix click/drag after scrolling

ajm188 commented on src/line.rs:
```diff
@@ -17,7 +16,7 @@ pub struct Line {
     #[serde(rename="cursor")]
     pub cursors: Option<Vec<u64>>,
-    pub styles: Option<Vec<i64>>,
+    pub styles: Vec<i64>,
```
is styles guaranteed to always be present? last I checked the protocol, that was not the case

little-dude replied:
In practice it seems to be the case but I'll ask on irc.
It makes the code more readable if we can get rid of `Option`, so that we don't have to write any access to styles with `if let Some(styles) = self.styles`/`self.styles.ok_or`/etc.

Same thing for the cursors btw.

little-dude replied:
Ok they're  _not_ guaranteed to be present:

```
<little_dude> morning everyone. small question: in the docs for updates (https://github.com/google/xi-editor/blob/master/doc/update.md) it is said that for a line, the "cursor" and "styles" fields are optional.
<little_dude> however in practice, they are always there, but are empty arrays
<raph> yes
<little_dude> so is it safe to assume they are always there?
<raph> hmm, for an "ins" they should in fact be missing, as a wire optimization
<raph> i thought i saw code for that
<raph> so i wouldn't assume they're always there
<raph> (though you can make a case that these little optimizations are not worth it)
<raph> keep in mind the rules for "update" are different than the rules for "ins"
<little_dude> each little piece of wood provides its little piece of fire
<raph> there, a missing cursor or styles means "unchanged"
<raph> common case, when you're moving the cursor, you'll update the "cursor" fields, nothing else
<little_dude> mmh I think I have this logic right already, but I'll double check.
<raph> we should have test vectors for it
<raph> especially as it's logic all of the front-ends have to get right
```

ajm188 commented on src/line.rs:
```diff
@@ -59,30 +57,42 @@ impl Line {
+            return Ok(());
+        }
+        if self.styles.len() % 3 != 0 {
+            // FIXME: find a more descriptive error for this. Hopefully, this should not happen
```
need a issue number here

ajm188 commented on src/line.rs:
```diff
@@ -59,30 +57,42 @@ impl Line {
-                    }
-                }
+            if end > text.len() {
+                // XXX: when dragging to a line below, xi-core sends a style that ends 1 index
```
add a TODO and open an issue?

little-dude replied:
After thinking more about this, I really think this is normal behaviour, and we just need to handle it.
I'll remove this comment.

Commit:
remove FIXME/XXX comments that do not apply

Commit:
use default values for Line fields

Line::cursors, Line::text and Line::styles are all optional fields, but
it's tedious to deal with Option<_> everywhere. Instead, use serde's
`default` attribute to set these fields to a default value when they are
missing.

little-dude commented:
I think this is ready :)
It's working fine for me.
